### PR TITLE
.github/workflows: move snap building to test.yaml as separate cached job

### PR DESCRIPTION
**Pull request by anonymouse64**
This will make it easier to use eventually within the spread test jobs and hopefully helps with some of the queuing we saw with using a separate workflow for this.

Note that after looking into a little bit, I'm not actually convinced that having this as a separate workflow that runs via the github hosted workers actually leads to longer queues, and I'm also not sure if moving this to the Tests workflow will make the job run on the self-hosted workers or if something manual needs to be done for this job to be run on the self-hosted workers.

### CODE REVIEW

**Comment by anonymouse64** [comment]
Hmm, upon moving the workflow here, the upload action completes successfully, but I can't actually see where the artifacts are the same way that I could with the job being in the other workflow ... needs some debugging

**Comment by anonymouse64** [comment]
> Hmm, upon moving the workflow here, the upload action completes successfully, but I can't actually see where the artifacts are the same way that I could with the job being in the other workflow ... needs some debugging

ah actually the artifact only shows up after all the tests have completed in the workflow, which is kinda unfortunate actually as sometimes the tests can be running for a long time, and the snap build finishes rather quickly after opening a PR :-/

**Comment by bboozzoo** [comment]
Pushed a fix, hopefully caching works now.

**Comment by anonymouse64** [comment]
The caching is successful now

**Comment by bboozzoo** [comment]
> The caching is successful now

However, there's no artifact after the latest run that was actually skipped because of cache. Perhaps the snap should also be part of the cache and in the build step, we should skip the actual build, but only move the upload the cached & restored snap to a location that the upload step will pick it up from.

**Comment by jhenstridge** [comment]
> I'm also not sure if moving this to the Tests workflow will make the job run on the self-hosted workers or if something manual needs to be done for this job to be run on the self-hosted workers.

This is dependent on the `runs-on` property of the job.  The `spread` and `spread-nested` jobs have `runs-on: self-hosted`, so run on Canonical's infrastructure.  The job you're adding has `runs-on: ubuntu-16.04`, so will run on Github's infrastructure.

You could probably also switch this job to `runs-on: ubuntu-latest` too.  The actual snap build will be performed within an LXD container selected by the `snapcraft.yaml` file, so there's no reason to also select the older VM image.

**Comment by bboozzoo** [comment]
With the latest fixes, the caching works and the artifact is uploaded after a cached run.

**Comment by anonymouse64** [comment]
> You could probably also switch this job to runs-on: ubuntu-latest too.

Thanks @jhenstridge for the suggestion, I think I will take this on a followup, since this PR is already green.

File: .github/workflows/test.yaml [hunk 1]
```diff
@@ -3,6 +3,51 @@ on:
...
+              has_cached_snap=0
+              while read name; do
+                  has_cached_snap=1
+                  # bring back artifacts from the cache
+                  cp -v "$name" "${{ github.workspace }}"
+              done < <(find "$(dirname $CACHE_RESULT_STAMP)" -name "*.snap")
+              if [ "$has_cached_snap" = "1" ]; then
+                  # we have restored an artifact from the cache
+                  echo "::set-output name=already-ran::true"
+              fi
```
Comment (anonymouse64) [inline .github/workflows/test.yaml:32]: heh, this ended up being quite a bit more complicated than I thought, glad you figured it out though

**Pull request merged**